### ISSUE TITLE
how to train multiple generations

### ISSUE THREAD

**Issue by anmolbyte** [DeepLabCut/DeepLabCut#1542]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Bug description

With the recent update, my old jupyter notebook for training -- which was working for multiple projects -- has started to not work. 

I used to:
1.) create a a new project and train (for this example lets say I only trained for 100 iterations)
2.) Analyze naïve videos 
3.)I determined the network had to be further trained
4.) extract_outlier_frames
5.) refine_labels
6.) merge_dataset
7.) create_new_training
8.) train_network 

What should the new proper function order be to ensure proper training of an i-generation dataset? I tried using add_new_videos (specifically the videos that I am extracting outlier frames for) before running the i generation of training.  This adds video paths to the actual config file, which was not something that was needed in previous versions. 

Additionally, it used to be that on the ith generation of training, the number of iterations was not 100*i, but instead, just 100, and the function understood that this meant 100 more. Additionally, this method now does not merge the training sets, but instead overwrites the existing one. 

### Operating System

Windows 10

### DeepLabCut version

2.2.0.2

### DeepLabCut mode

single animal

### Device type

1050TI

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/DeepLabCut/DeepLabCut/blob/master/CODE_OF_CONDUCT.md)

**Comment by backyardbiomech** [comment]
With the actual error, it's impossible to tell what you mean by "has started not to work."

Videos used in training always had to be added to the config file. The paths to the videos is how deeplabcut knows from which folders to pull images and data during training set creation. If you don't have a path to a video in config, even if you do have a labeled data folder for that video, it won't be included (that's an easy way to temporarily eliminate data from a specific video from training). Extracting outliers from novel videos, however, adds the videos to config, so chances are they were always being added without you realizing it.

But, yes, the way that snapshots are counted, when training from a previous snapshot, was updated with DLC 2.2. Now, the snapshot iteration counter continues from the previous iteration, but only when starting from a previous snapshot. If you start a new round from the default network, your iterations will start at 0.  Note that when starting from a previous snapshot, you now also have to update your multi-step learning stages and maxiters, because those take into account the new iteration count. I.e., if you have `maxiters=20000`, and you are starting from a previous snapshot of 20000, training will stop immediately because it will hit the max as soon as it opens. 

In my experience, I very rarely start from a previous snapshot. I only do it when I have a very accurate model, but have some new view or behavior that it just can't quite figure out. Otherwise, I always start from scratch - it takes a bit more time, but the end product is usually more accurate that many short training sets from previous snapshots.

**Comment by MMathisLab** [comment]
"I tried using add_new_videos (specifically the videos that I am extracting outlier frames for) before running the i generation of training. This adds video paths to the actual config file, which was not something that was needed in previous versions."

-- this was always the behavior of deeplabcut. 

In short, yes the beta versions are *not* forward compatible, as we explained in detail in the release notes and blog: blog.deeplabcut.org 

- you should use the latest notebooks, especially the colab ones have a lot of details / and see the docs: https://deeplabcut.github.io/DeepLabCut/docs/intro.html

**Comment by MMathisLab** [comment]
"Additionally, this method now does not merge the training sets, but instead overwrites the existing one." --> can you clarify, i.e. you run merge_datasets and you are saying that old data is dropped and only outlier are there?

**Comment by anmolbyte** [comment]
@backyardbiomech  To clarify, I would extract_outlier_frames from analyzed videos. Where did these extracted labeled frames get used? Were they not used to train if I did not add the associated videos using the add_new_videos to the config file ? When I look at the config file, and didnt run add_new_videos, I don't see the new videos that I extracted outlier frames from. Additionally  the merge_dataset over-wrote the labeled data used for training.

Also, thanks for letting me know about the new iteration update!

Also, why would you start from scratch and not update the existing weights? So you suggest only using 1 generation of data? It is to my understanding that it is better to use a lot of training videos with little labeled frames compared to one training video and a lot of labeled frames, although how the number of generations effects this process is unknown to me.
 

@MMathisLab Sorry for missing the that part of the new update notes. 

As for the clarification (@backyardbiomech, this is related to your question too), I usually extract outlier frames and then merge the data sets, which I thought took the new extracted labeled frames and merge them with the old training set, and added their associated videoes to the config file. But, Iit looked like when I was running the newest version of DLC, the function merge_Dataset and extract_outlier_codes did not add the videos to the config file yet the extracted frames were overwriting the previous generation's labeled frames for training. In other words it looked like the training data was not even for the correct training video specified in the config file.

To summarize and reiterate, I want a way to properly generate two+ generations of training, which I have seemed to do incorrectly. What is the proper order of operations for this? 

Does this make sense? Thanks a ton for the help and the patience!

**Comment by backyardbiomech** [comment]
@ajarang : Here's how it's supposed to work: After analyzing a video, you run `extract_outlier_frames`. 

**If the video was one that you already added and labeled**, new frames would be extracted and saved in `projectfolder/labeled-data/vidname`, and the coordinates of the outliers saved in `machinelabels-iterx.h5` in that  folder. That's the file you open in the `refine_labels` GUI. Once those are refined, when you click `Save` the coordinates get moved to the `CollectedData-DLC.h5` and `csv` file, without overwriting the previous labeled frames there.
After all that, merging the data mostly just increases the `iteration` parameter in `config.yaml` by 1.

**If the video has not yet been added to the project and labeled**, then the code tries to add the the path of the video to `config.yaml`. You *might* get an error in the terminal like `"AUTOMATIC ADDING OF VIDEO TO CONFIG FILE FAILED! You need to do this manually for including it in the config.yaml file!"`, in which case you need to manually add the video. Note that this will **not** copy the video to the videos folder, so if you like to have the videos there (or can't create shortcuts due to permissions issues), then copy the video and tracked coordinate files there before running extract outliers. The extract outliers code then creates a folder for the video in `labeled-data` using the name of the video, and saves the extracted frame images, the coordinates in `machinelabels-iterx.h5`, and creates a CollecteData file. 

So, you shouldn't need to run `add_new_videos` since that is called by `extract_outlier_frames` [on line 670](https://github.com/DeepLabCut/DeepLabCut/blob/2c5f2a76d772596946654ba15a6e7f35f08937b0/deeplabcut/refine_training_dataset/outlier_frames.py#L670). 

It should also not overwrite previous data.

The one thing I can think of is that maybe your video files have the same name? (e.g., `cam1.mov`). Since the folders in `labeled-data` are based on the video file names, each video used in building the training set must have unique names. If you don't have unique names, then I could see where some very odd behavior would creep in.

**Comment by hummuscience** [comment]
"After all that, merging the data mostly just increases the iteration parameter in config.yaml by 1."

This does not happen in the current version (using single animal DLC)

**Comment by MMathisLab** [comment]
@Cumol - it definitely does. You can run the testscript.py if any doubts; there it tests extracting outliers, labeling, `merge_datasets`... and you can see it bumps up config.yaml to 1:

https://github.com/DeepLabCut/DeepLabCut/blob/66d483924f4746ee2f8362303901ca2e0931a3c0/examples/testscript.py#L251

```python
Successfully installed deeplabcut-2.2.0.2 opencv-python-4.5.4.58 scikit-image-0.18.1
❯ cd examples
❯ pythonw testscript.py
...
 'snapshot_prefix': '/Users/mwmathis/Documents/DeepLabCut/examples/TEST-Alex-2021-10-25/dlc-models/iteration-1/TESTOct25-trainset80shuffle2/train/snapshot',
 'stride': 8.0,
 'weigh_negatives': False,
 'weigh_only_present_joints': False,
 'weigh_part_predictions': False,
 'weight_decay': 0.0001}
Merging datasets...
Creating two identical splits...
You passed a split with the following fraction: 80%
You passed a split with the following fraction: 80%
The training dataset is successfully created. Use the function 'train_network' to start training. Happy training!
The training dataset is successfully created. Use the function 'train_network' to start training. Happy training!
ALL DONE!!! - default cases are functional.
```

Image of my config.yaml file in the test project:
<img width="392" alt="Screen Shot 2021-10-25 at 4 13 02 PM" src="https://user-images.githubusercontent.com/28102185/138712153-ecb0d257-e063-4b7c-91f3-88d0f1f96f43.png">

**Comment by anmolbyte** [comment]
Awesome. This is quite helpful. 

I have also noticed something about DLC_3d:
-I love how the new update will automatically take away frames that did not pass the find grid function, but I noticed that sometimes it will remove a frame for one camera (e.g. cam_1-frame 10) and then before even running the find grid function, it will remove the associated frame (e.g. cam_2 -fram 10) for the other camera. 

It seems that sometimes the find-grid still "looks" for the frame that was removed and so I get an error (usually a cv2.RBG2RAY error) because it is looking for the empty frame. I've also noticed that just rerunning the calibrate_cameras function a few times will make this error eventually go away as if removes the frames and stops searching. 

Any suggestions to do this without getting the error? 

Y'all are the best!

**Comment by anmolbyte** [comment]
additionally, my calibrate_cameras is taking over 5 hours and is still not outputting the 3 files.

any ideas?
